src/astro/utils/table_handler.py

```diff
@@ -16,28 +16,50 @@ def _set_variables_from_first_table(self):
         """
         first_table: Optional[Table] = None
         if self.op_args:
-            table_index = [x for x, t in enumerate(self.op_args) if type(t) == Table]
-            if table_index:
+            table_index = [
+                x for x, t in enumerate(self.op_args) if isinstance(t, Table)
+            ]
+            conn_id_set = {x.conn_id for x in self.op_args if isinstance(x, Table)}
+            # Check to see if all tables belong to same conn_id. Otherwise, we this can go wrong for cases
+            # 1. When we have tables from different DBs.
+            # 2. When we have tables from different conn_id, since they can be configured with different
+            # database/schema etc.
+            if table_index and len(conn_id_set) == 1:
                 first_table = self.op_args[table_index[0]]
         elif not first_table:
             table_kwargs = [
                 x
                 for x in inspect.signature(self.python_callable).parameters.values()
                 if (
                     x.annotation == Table
-                    and type(self.op_kwargs[x.name]) == Table
+                    and isinstance(self.op_kwargs[x.name], Table)
                     or x.annotation == pandas.DataFrame
-                    and type(self.op_kwargs[x.name]) == Table
+                    and isinstance(self.op_kwargs[x.name], Table)
                 )
             ]
-            if table_kwargs:
+            conn_id_set = {
+                self.op_kwargs[x.name].conn_id
+                for x in inspect.signature(self.python_callable).parameters.values()
+                if (
+                    x.annotation == Table
+                    and isinstance(self.op_kwargs[x.name], Table)
+                    or x.annotation == pandas.DataFrame
+                    and isinstance(self.op_kwargs[x.name], Table)
+                )
+            }
+            if table_kwargs and len(conn_id_set) == 1:
                 first_table = self.op_kwargs[table_kwargs[0].name]
 
         # If there is no first table via op_ags or kwargs, we check the parameters
         elif not first_table:
             if self.parameters:
-                param_tables = [t for t in self.parameters.values() if type(t) == Table]
-                if param_tables:
+                param_tables = [
+                    t for t in self.parameters.values() if isinstance(t, Table)
+                ]
+                conn_id_set = {
+                    t.conn_id for t in self.parameters.values() if isinstance(t, Table)
+                }
+                if param_tables and len(conn_id_set) == 1:
```

Comment:
@utkarsharma2 as we spoke, since this was a bug found in an Astro release, we need a regression test to make sure the same bug doesn't appear again - particularly given the refactoring we're doing.

Reply:
It should be possible to unit test `_set_variables_from_first_table`

Reply:
I have added a UnitTest case checking all possible conditions.